vim-patch:8.1.{1631,1682,1899}: sign improvements

lewis6991 opened the pull request:
### vim-patch:8.1.1631: displaying signs is inefficient
Original PR: https://github.com/vim/vim/pull/4586

> When displaying a sign, multiple calls are made to get information
related to a sign. Each call walks through all the signs placed
in the buffer. Add a function to get all the sign attributes in a single
call and use it.

**Note:** This patch is incompatibilites with Neovim's dynamic signcolumn. I have adapted the code as best I can to achieve the same effect of the patch by calcuating the sign attributes once for all signs on the line.

### vim-patch:8.1.1682: placing a larger number of signs is too slow
Original PR: https://github.com/vim/vim/pull/4636

> vim-patch:8.1.1682: placing a larger number of signs is too slow
> Add support for:
> - defining multiple signs using sign_define()
> - undefining multiple signs using sign_undefine()
> - placing multiple signs using sign_placelist()
> - removing multiple signs using sign_unplacelist()

### vim-patch:8.1.1899: sign_place() does not work as documented 
Original PR: https://github.com/vim/vim/pull/4848

> The help text for sign_place() states that the lnum argument accepts values
that are supported by line(). Modify the code to make it so.

lewis6991 committed:
vim-patch:8.1.1682: placing a larger number of ...

...signs is slow

Problem:    Placing a larger number of signs is slow.
Solution:   Add functions for dealing with a list of signs. (Yegappan
            Lakshmanan, closes #4636)

lewis6991 committed:
vim-patch:8.1.1899: sign_place() does not work as documented

Problem:    sign_place() does not work as documented.
Solution:   Make accept line numbers like line(). (Yegappan Lakshmanan,
            closes #4848)

janlazo commented on src/nvim/testdir/otherName:
```diff
@@ -0,0 +1,2 @@
+some
```
Delete

janlazo commented on src/nvim/testdir/test_result.log:
```diff
@@ -0,0 +1,147 @@
+
```
Delete

janlazo commented on src/nvim/version.c:
```diff
@@ -67,6 +67,7 @@ NULL
 
 // clang-format off
 static const int included_patches[] = {
+  1899,
```
Revert.

These patch numbers target Vim 8.0. Neovim has patches across multiple Vim versions so `included_patches` updates are automated.

janlazo commented on src/nvim/sign.c:
```diff
@@ -974,7 +977,7 @@ int sign_place(
         has_text_or_icon);
   } else {
     // ":sign place {id} file={fname}": change sign type
```
Vim changed this comment.

janlazo commented on src/nvim/sign.c:
```diff
@@ -1860,3 +1863,267 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+    }
+    name = tv_dict_get_string(dict, "name", true);
+  } else {
+    name = (char *)vim_strsave((char_u *)name_arg);
```
Use `xstrdup` directly.

janlazo commented on src/nvim/testdir/test_signs.vim:
```diff
@@ -1143,7 +1153,7 @@ func Test_sign_id_autogen()
   edit Xsign
 
-  call assert_equal(1, sign_place(0, '', 'sign1', 'Xsign',
+  call assert_equal(2, sign_place(0, '', 'sign1', 'Xsign',
```
Bug?

lewis6991 replied:
Still need to investigate whether this is ok. As it is only an id I don't think it needs to be identical across versions of the code, it just need to be unique.

janlazo commented on src/nvim/testdir/test_signs.vim:
```diff
@@ -1159,8 +1169,8 @@ func Test_sign_id_autogen()
   call assert_equal(0, sign_unplace('g1', {'id' : 1}))
-  call assert_equal(10,
-	      \ sign_getplaced('Xsign', {'id' : 1})[0].signs[0].lnum)
+  call assert_equal(12,
```
Bug?

lewis6991 replied:
As above

janlazo commented on src/nvim/testdir/test_signs.vim:
```diff
@@ -1869,3 +1879,121 @@ func Test_sign_numcol()
+  call sign_unplace('*')
+
+  " Place multiple signs at once with auto-generated sign identifier
+  call assert_equal([1, 1, 1], sign_placelist([
```
[1,1,5]

lewis6991 replied:
Again these are only id's so I don't think we need to have the same results as vim. The test placeing signs for three different groups so it's ok that the signs all have the same id.

janlazo reviewed (changes requested):
There are extra changes in test_signs.vim which may suggest missing patches or undocumented difference between Vim and Neovim.

Please fix the minor issues I identified. The rest looks fine.

lewis6991 committed:
vim-patch:8.1.1631: displaying signs is inefficient

+ support for neovim's dynamic width signcolumn

Problem:    Displaying signs is inefficient.
Solution:   Avoid making multiple calls to get information about a placed
            sign. (Yegappan Lakshmanan, closes #4586)

lewis6991 commented:
Status update on this.

I ended up also adding vim patch 8.1.1631 which provides performance improvements by calculating sign attributes ahead of time instead of doing multiple of lookups for each, which looking at it is really inefificent. Because of neovim's multi-signcolumn feature I had to significantly rework this patch so we calculate upto 9 sets of sign attrs instead of 1.

The test discrepency in the last review was due to test ordering. I omitted part of the original patch which added `aaa` into a test name to force it to run before other tests. After adding this, the test changes could be reverted to match vim.

Other than that I think I've addressed all the previous review comments but a may have missed one.

Thanks.

janlazo reviewed (approved):
LGTM
